New module for azure custom image facts

Madhura-CSI opened the pull request:
##### SUMMARY
New module to collect facts about azure custom images.
<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 
 - New Module Pull Request


##### COMPONENT NAME
azure_rm_image_facts

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.4
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/centos-dev/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```


##### ADDITIONAL INFORMATION
Currently we have module to gather information about Virtual Machine Images based on location, sku, offer, etc. But we don't seem to have module to collect information about custom images created from virtual machines. This module helps us in retrieving those data.

Module Details:
It takes 2 parameters(both optional)
1. resource_group
2. Name

Flow of module execution:
1. If name and resource group both are specified, it gets information about particular image in a resource group.
2. If only name is specified, gets information about all matching images with 'name' in a current subscription.
3. If only resource group is specified, gets information about all custom images within a resource group.
4. If no parameter is specified, gets information about all custom images in a current subscription.

Sample module invocations are:
<!--- Paste verbatim command output below, e.g. before and after your change -->
```
- name: List images with name
  azure_rm_image_facts:
    name: test-image
    resource_group: test-resource-group

- name: List all available images under current subscription
  azure_rm_image_facts:
  register: all_images
```
I would like to contribute this module to community.

Madhura-CSI committed:
new module for azure custom image facts

yuwzho commented:
@Madhura-CSI Thanks for your contribution, these days we are trying to make some change for facts module's return value. We don't want to simply return the serialization JSON, but return a dict for each item, the dict should contain information  which is the same as parameters required by resource module, together with a resource id. That means, the  `azure_image_facts` module should return a list like this:

```json
{
  "azure_images": [
    {
      "id": "xxxxx",
      "name": "XXXX",
      "source": "XXXX",
      "data_disk_sources": "XXXX",
      "os_type": "XXXX",
      "tags": {
        "xxx": "xxx"
      }
    }
  ]
}
```

yungezz commented:
thanks for contribution. we're trying to generalize facts module return format. agreement has been reached to support `curated` output which will match input to resource module. @zikalino  could you pls share `curated` and `raw` option definition for reference?

yuwzho commented:
duplicate_of #32403

Fred-sun commented:
@Madhura-CSI Thanks for your contribution, Are you still here? Could you help update the PR status? How to make it step forward? Thanks!

Madhura-CSI commented:
@Fred-sun , started working on review comments i.e. to make use of ```curated``` and ```raw``` formats. Will update the code once done with the changes.

Fred-sun commented:
@Madhura-CSI Thanks for you update comment, after you finish change, I will push to review. Thanks!

Fred-sun commented:
@Madhura-CSI Thanks for the contribute, Could you help confirm that you have finish change? Thanks!

Madhura-CSI commented:
@Fred-sun @yungezz @yuwzho , I have addressed review comment for the module and also have updated documentation section with example.
Please review and let me know if you would like to have further modifications done.

Madhura-CSI commented:
ready_for_review

Fred-sun commented:
@Madhura-CSI Thanks for the contribute.

Fred-sun commented:
@zikalino @yungezz @yuwzho This PR ready for review, Could you help review it? Thanks!

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_image_facts.py:
```diff
@@ -0,0 +1,268 @@
+    tags:
+        description:
+            - List of tags to be matched.
+    format:
```
For the new added facts, we suggests no `raw`

Madhura-CSI replied:
Please correct me if I am wrong, so does this mean we should remove support for ```raw``` and keep only curated format? (something similar like https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/azure/azure_rm_postgresqldatabase_facts.py ?)

yuwzho replied:
Yes, it is. Some modules contain both raw and curated because of the compatible issue

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_image_facts.py:
```diff
@@ -0,0 +1,268 @@
+
+- name: List all available images under current subscription
+  azure_rm_image_facts:
+  register: all_images
```
remove the register line, kind of confused

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_image_facts.py:
```diff
@@ -0,0 +1,268 @@
+                    data_disks[disk_index] = new_disk
+
+                updated = {
+                    'resource_group': self.resource_group,
```
resource group should be parsed from image.id, considering the list_all scenario.

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_image_facts.py:
```diff
@@ -0,0 +1,268 @@
+
+        if self.format == 'curated':
+            for index in range(len(self.results['ansible_facts']['azure_images'])):
+                image = self.results['ansible_facts']['azure_images'][index]
```
I would like to use a more python way: list comprehension and a `to_disk` method to construct the list.

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_image_facts.py:
```diff
@@ -0,0 +1,268 @@
+        if response:
+            for item in response:
+                if self.has_tags(item.tags, self.tags):
+                    results.append(self.serialize_obj(item, 'ImagePaged',
```
IMO, there is no need to serialize the object if user choose a curated way. So we can only keep the tag filter here as 
```python
result = [x for x in response if self.has_tags(x.tags, self.tags)] if response else []
```

Fred-sun commented:
@Madhura-CSI Thanks for the contribution, Could you help update the comment according by @yuwzho's advice? Thanks!

Madhura-CSI commented:
ready_for_review

Madhura-CSI commented:
ready_for_review

Fred-sun commented:
@yuwzho @zikalino Could you help to review this PR？I have test this function and get successfully result. Thanks!

yuwzho commented:
Would you mind adding a test for facts module int he https://github.com/ansible/ansible/blob/devel/test/integration/targets/azure_rm_image/tasks/main.yml ?

yuwzho commented:
shipit

yuwzho commented:
It fails with following message according to https://app.shippable.com/github/ansible/ansible/runs/81386/74/console


```text
07:54 TASK [azure_rm_image : Gather information about image created] *****************
07:55 fatal: [localhost]: FAILED! => {
  "changed": false,
  "module_stderr": "No handlers could be found for logger \"msrestazure.azure_active_directory\"
			Traceback (most recent call last):
			  File \"/root/.ansible/tmp/ansible-tmp-1535523945.75-76270332090746/AnsiballZ_azure_rm_image_facts.py\", line 113, in <module>
			    _ansiballz_main()
			  File \"/root/.ansible/tmp/ansible-tmp-1535523945.75-76270332090746/AnsiballZ_azure_rm_image_facts.py\", line 105, in _ansiballz_main
			    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
			  File \"/root/.ansible/tmp/ansible-tmp-1535523945.75-76270332090746/AnsiballZ_azure_rm_image_facts.py\", line 48, in invoke_module
			    imp.load_module('__main__', mod, module, MOD_DESC)
			  File \"/tmp/ansible_azure_rm_image_facts_payload_tb9QNk/__main__.py\", line 231, in <module>
			  File \"/tmp/ansible_azure_rm_image_facts_payload_tb9QNk/__main__.py\", line 227, in main
			  File \"/tmp/ansible_azure_rm_image_facts_payload_tb9QNk/__main__.py\", line 140, in __init__
			  File \"/tmp/ansible_azure_rm_image_facts_payload_tb9QNk/ansible_azure_rm_image_facts_payload.zip/ansible/module_utils/azure_rm_common.py\", line 391, in __init__
			  File \"/tmp/ansible_azure_rm_image_facts_payload_tb9QNk/__main__.py\", line 149, in exec_module
			  File \"/tmp/ansible_azure_rm_image_facts_payload_tb9QNk/__main__.py\", line 172, in get_image
			  File \"/tmp/ansible_azure_rm_image_facts_payload_tb9QNk/__main__.py\", line 219, in format_item
			KeyError: 'managed_disk'",
  "module_stdout": "",
  "msg": "MODULE FAILURE
  See stdout/stderr for the exact error",
  "rc": 1
}
```

Madhura-CSI commented:
@yuwzho, I have addressed the error you mentioned above in latest commit. Please review.
Thanks!

yuwzho commented:
shipit

@yungezz @zikalino

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_image_facts.py:
```diff
@@ -0,0 +1,232 @@
+            'os_state': d['storage_profile']['os_disk']['os_state'],
+            'os_disk_caching': d['storage_profile']['os_disk']['caching'],
+            'os_storage_account_type': d['storage_profile']['os_disk']['storage_account_type'],
+            'os_disk': d['storage_profile']['os_disk']['managed_disk']['id'] if 'managed_disk' in d['storage_profile']['os_disk'].keys() else None,
```
Also as the `os_disk_caching` and other options?

yuwzho commented on lib/ansible/modules/cloud/azure/azure_rm_image_facts.py:
```diff
@@ -0,0 +1,232 @@
+            'os_state': d['storage_profile']['os_disk']['os_state'],
+            'os_disk_caching': d['storage_profile']['os_disk']['caching'],
+            'os_storage_account_type': d['storage_profile']['os_disk']['storage_account_type'],
+            'os_disk': d['storage_profile']['os_disk']['managed_disk']['id'] if 'managed_disk' in d['storage_profile']['os_disk'].keys() else None,
```
Also as the os_disk_caching and other options?

Madhura-CSI replied:
It seems os_disk_caching variable value is always set in response, from [None, ReadyOnly, ReadWrite]. So I think it is okay to skip the condition on it.

yuwzho commented:
shipit

yungezz commented on lib/ansible/modules/cloud/azure/azure_rm_image_facts.py:
```diff
@@ -0,0 +1,232 @@
+
+        self.results = dict(
+            changed=False,
+            ansible_facts=dict(azure_images=[])
```
pls remove ansible_facts, and azure from the  result, just `images=[]`. we're working on general code guideline on facts module, one of them is shorten the result level, rule is no ansible_facts , and azure key word, since it's already  azure module.

Madhura-CSI replied:
Please check the latest commit. I have removed 'azure' from resultant list name and removed the same from ansible_facts as well.

yungezz commented:
@zikalino

zikalino commented on lib/ansible/modules/cloud/azure/azure_rm_image_facts.py:
```diff
@@ -0,0 +1,231 @@
+    description: List of image dicts.
+    returned: always
+    type: list
+    example: [{
```
instead of example now we need proper description of returned values, please check facts modules like azure_rm_webapp_facts or azure_rm_containerregistry_facts as a reference

zikalino commented on test/integration/targets/azure_rm_image/tasks/main.yml:
```diff
@@ -116,6 +116,16 @@
+    name: testimage001
+  register: output
+
+- assert:
```
please also add "azure_rm_image_facts" alias in aliases

zikalino reviewed (changes requested):
@Madhura-CSI pls just update return value description, and then i will merge. everything else looks very good!

Madhura-CSI committed:
added facts in aliases, corrected assertion

Madhura-CSI commented:
@zikalino , I have updated examples section of module and also corrected assertion in integration tests.
Latest ansibot run is successful. Please review.
Thanks!

zikalino commented on lib/ansible/modules/cloud/azure/azure_rm_image_facts.py:
```diff
@@ -0,0 +1,300 @@
+                        - Specifies the storage account type for the managed disk data disk.
+                    type: str
+                    sample: Standard_LRS
+                managed_disk:
```
could we just have managed_disk_id? seems like a structure with a single field...

Madhura-CSI replied:
Agree with you @zikalino . But it seems it can be an overhead with further processing, by iterating over all data_disks. Please let me know if it is feasible to do this.

zikalino replied:
yes, it's ok to iterate and change, as long as users will get simpler structure :-)

Madhura-CSI replied:
done with the changes. Please review @zikalino .
Thanks!

yungezz commented:
shipit

Madhura-CSI committed:
simplified managed disk id struct

Madhura-CSI committed:
updated ansible version

zikalino commented:
@Madhura-CSI great!
